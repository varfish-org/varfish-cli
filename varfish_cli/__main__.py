"""Main entry point for VarFish CLI."""

import argparse
import logging
import os
import sys

import logzero
from logzero import logger
import toml

from varfish_cli import __version__
from varfish_cli.case import run as run_case
from varfish_cli.case import setup_argparse as setup_argparse_case
from varfish_cli.common import CommonConfig, run_nocmd
from varfish_cli.varannos import run as run_varannos
from varfish_cli.varannos import setup_argparse as setup_argparse_varannos

#: Paths to search the global configuration in.
GLOBAL_CONFIG_PATHS = ("~/.varfishrc.toml",)


def setup_argparse_only():  # pragma: nocover
    """Wrapper for ``setup_argparse()`` that only returns the parser.

    Only used in sphinx documentation via ``sphinx-argparse``.
    """
    return setup_argparse()[0]


def setup_argparse():
    """Create argument parser."""
    # Construct argument parser and set global options.
    parser = argparse.ArgumentParser(prog="varfish-cli")
    parser.add_argument("--verbose", action="store_true", default=False, help="Increase verbosity.")
    parser.add_argument("--version", action="version", version="%%(prog)s %s" % __version__)

    group = parser.add_argument_group("Basic Configuration")
    group.add_argument(
        "--no-verify-ssl",
        dest="verify_ssl",
        default=True,
        action="store_false",
        help="Disable HTTPS SSL verification",
    )
    group.add_argument(
        "--config",
        default=os.environ.get("VARFISH_CONFIG_PATH", None),
        help="Path to configuration file.",
    )
    group.add_argument(
        "--varfish-server-url",
        default=os.environ.get("VARFISH_SERVER_URL", None),
        help="VarFish server URL key to use, defaults to env VARFISH_SERVER_URL.",
    )
    group.add_argument(
        "--varfish-api-token",
        default=os.environ.get("VARFISH_API_TOKEN", None),
        help="VarFish API token to use, defaults to env VARFISH_API_TOKEN.",
    )

    # Add sub parsers for each argument.
    subparsers = parser.add_subparsers(dest="cmd")

    setup_argparse_case(subparsers.add_parser("case", help="Work with cases."))
    setup_argparse_varannos(subparsers.add_parser("varannos", help="Work with varannos module."))

    return parser, subparsers


def main(argv=None):
    """Main entry point before parsing command line arguments."""
    # Setup command line parser.
    parser, subparsers = setup_argparse()

    # Actually parse command line arguments.
    args = parser.parse_args(argv)

    # Setup logging incl. verbosity.
    if args.verbose:  # pragma: no cover
        level = logging.DEBUG
    else:
        # Remove module name and line number if not running in debug mode.s
        formatter = logzero.LogFormatter(
            fmt="%(color)s[%(levelname)1.1s %(asctime)s]%(end_color)s %(message)s"
        )
        logzero.formatter(formatter)
        level = logging.INFO
    logzero.loglevel(level=level)

    # Load configuration, if any.
    if args.config:
        config_paths = (args.config,)
    else:
        config_paths = GLOBAL_CONFIG_PATHS
    for config_path in config_paths:
        config_path = os.path.expanduser(os.path.expandvars(config_path))
        if os.path.exists(config_path):
            with open(config_path, "rt") as tomlf:
                toml_config = toml.load(tomlf)
            break
    else:
        toml_config = None
        logger.info("Could not find any of the global configuration files %s.", config_paths)

    # Merge configuration from command line/environment args and configuration file.
    config = CommonConfig.create(args, toml_config)

    # Handle the actual command line.
    cmds = {None: run_nocmd, "case": run_case, "varannos": run_varannos}

    res = cmds[args.cmd](
        config, toml_config, args, parser, subparsers.choices[args.cmd] if args.cmd else None
    )
    if not res:
        logger.info("All done. Have a nice day!")
    else:  # pragma: nocover
        logger.error("Something did not work out correctly.")
    return res


if __name__ == "__main__":  # pragma: no cover
    sys.exit(main(sys.argv))
