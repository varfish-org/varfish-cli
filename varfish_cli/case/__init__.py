"""Implementation of varfish-cli subcommand 'case'."""

import argparse

from varfish_cli.case.config import CaseConfig, OutputFormat
from varfish_cli.case.create_case_import import (
    setup_argparse as setup_argparse_create_import,
)
from varfish_cli.case.list_case import setup_argparse as setup_argparse_list
from varfish_cli.case.list_case_import import (
    setup_argparse as setup_argparse_list_import,
)
from varfish_cli.case.small_var_query_create import (
    setup_argparse as setup_argparse_small_var_query_create,
)
from varfish_cli.case.small_var_query_fetch_results import (
    setup_argparse as setup_argparse_small_var_query_fetch_results,
)
from varfish_cli.case.small_var_query_list import (
    setup_argparse as setup_argparse_small_var_query_list,
)
from varfish_cli.case.small_var_query_retrieve import (
    setup_argparse as setup_argparse_small_var_query_retrieve,
)
from varfish_cli.case.small_var_query_settings_shortcut import (
    setup_argparse as setup_argparse_small_var_query_settings_shortcut,
)
from varfish_cli.case.small_var_query_status import (
    setup_argparse as setup_argparse_small_var_query_status,
)
from varfish_cli.case.small_var_query_update import (
    setup_argparse as setup_argparse_small_var_query_update,
)
from varfish_cli.common import run_nocmd


def setup_argparse(parser: argparse.ArgumentParser) -> None:
    """Main entry point for subcommand."""
    of_choices = [o.value for o in OutputFormat]
    of_default = OutputFormat.TABLE.value
    parser.add_argument(
        "--output-format",
        help=f"Output format, one of {of_choices}, default: {of_default}",
        choices=of_choices,
        default=of_default,
    )
    parser.add_argument("--output-fields", help="Output fields if non-default ones")
    parser.add_argument(
        "--output-delimiter", help="Separator for CSV output, default: ','", default=","
    )
    parser.add_argument(
        "--output-file", help="Path to file to write to, defaults to stdout", default="-"
    )

    subparsers = parser.add_subparsers(dest="case_cmd")
    setup_argparse_list(subparsers.add_parser("list", help="List cases."))
    setup_argparse_list_import(
        subparsers.add_parser("list-import-info", help="List case import infos.")
    )
    setup_argparse_create_import(
        subparsers.add_parser("create-import-info", help="Create case import infos.")
    )
    setup_argparse_small_var_query_create(
        subparsers.add_parser("small-var-query-create", help="Create small var query")
    )
    setup_argparse_small_var_query_fetch_results(
        subparsers.add_parser("small-var-query-fetch-results", help="Fetch small var query results")
    )
    setup_argparse_small_var_query_list(
        subparsers.add_parser("small-var-query-list", help="List small var queries")
    )
    setup_argparse_small_var_query_retrieve(
        subparsers.add_parser("small-var-query-retrieve", help="Retrieve small variant query")
    )
    setup_argparse_small_var_query_status(
        subparsers.add_parser("small-var-query-status", help="Get small variant query status")
    )
    setup_argparse_small_var_query_update(
        subparsers.add_parser("small-var-query-update", help="Update small variant query")
    )
    setup_argparse_small_var_query_settings_shortcut(
        subparsers.add_parser(
            "small-var-query-shortcut", help="Generate query parameters from shortcut"
        )
    )


def run(config, toml_config, args, parser, subparser):
    """Main entry point for case command."""
    if not args.case_cmd:  # pragma: nocover
        return run_nocmd(config, args, parser, subparser)
    else:
        config = CaseConfig.create(args, config, toml_config)
        return args.case_cmd(config, toml_config, args, parser, subparser)
