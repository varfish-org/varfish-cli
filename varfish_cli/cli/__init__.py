"""Code for main entry point."""

import logging
import typing

import logzero
import typer

from varfish_cli import __version__
from varfish_cli.cli import cases, importer, projects, tools, varannos
from varfish_cli.config import CommonOptions, load_config
from varfish_cli.exceptions import InvalidConfiguration

#: Paths to search the global configuration in.
DEFAULT_PATH_VARFISHRC = "~/.varfishrc.toml"


def version_callback(value: bool):
    """Callback when called with 'version' or '--version'"""
    if value:
        print(f"varfish-cli {__version__}")
        raise typer.Exit()


#: Main CLI ``Typer`` object.
app = typer.Typer(no_args_is_help=True, pretty_exceptions_enable=False)

# Register all sub commands.
app.add_typer(varannos.app, name="varannos", help="Subcommands for 'varannos' API")
app.add_typer(projects.app, name="projects", help="Subcommands for 'project' API")
app.add_typer(cases.app, name="cases", help="Subcommands for 'cases' API")
app.add_typer(importer.app, name="importer", help="Subcommands for 'importer' API")
app.add_typer(tools.app, name="tools", help="Subcommands for 'tools' API")


@app.command("version")
def main_version():
    """Print version as "varfish-cli $version"."""
    version_callback(True)


@app.callback()
def main(
    ctx: typer.Context,
    version: typing.Annotated[
        typing.Optional[bool], typer.Option("--version", callback=version_callback)
    ] = None,
    verbose: typing.Annotated[
        bool, typer.Option("--verbose", "-v", help="Enable verbose output")
    ] = False,
    verify_ssl: typing.Annotated[
        bool, typer.Option("--verify-ssl/--no-verify-ssl", help="Disable SSL verification")
    ] = True,
    config_path: typing.Annotated[
        str,
        typer.Option("--config-path", help="Path to configuration file", envvar="VARFISH_RC_PATH"),
    ] = DEFAULT_PATH_VARFISHRC,
    varfish_server_url: typing.Annotated[
        typing.Optional[str],
        typer.Option(
            "--varfish-server-url",
            help=(
                "VarFish server URL key to use, defaults to env VARFISH_SERVER_URL or read "
                "from configfile"
            ),
            envvar="VARFISH_SERVER_URL",
        ),
    ] = None,
    varfish_api_token: typing.Annotated[
        typing.Optional[str],
        typer.Option(
            "--varfish-server-url",
            help=(
                "VarFish API token to use, defaults to env VARFISH_API_TOKEN or read from "
                "configfile"
            ),
            envvar="VARFISH_API_TOKEN",
        ),
    ] = None,
):
    """Callback for main entry point

    This function handles the global configuration from configuration file,
    environment variables, and command line (in increasing priority).
    """
    _ = version
    # Setup logging
    if verbose:  # pragma: no cover
        level = logging.DEBUG
    else:
        # Remove module name and line number if not running in debug mode.s
        formatter = logzero.LogFormatter(
            fmt="%(color)s[%(levelname)1.1s %(asctime)s]%(end_color)s %(message)s"
        )
        logzero.formatter(formatter)
        level = logging.INFO
    logzero.loglevel(level=level)

    # Load configuration file
    toml_varfish_server_url, toml_varfish_api_token = load_config(config_path)
    if toml_varfish_server_url and not varfish_server_url:
        varfish_server_url = toml_varfish_server_url
    if toml_varfish_api_token and not varfish_api_token:
        varfish_api_token = toml_varfish_api_token

    if not varfish_server_url or not varfish_api_token:
        raise InvalidConfiguration("Need to specify server URL and API token")

    # Construct common options
    ctx.obj = CommonOptions(
        verbose=verbose,
        verify_ssl=verify_ssl,
        config_path=config_path,
        varfish_server_url=varfish_server_url,
        varfish_api_token=varfish_api_token,
    )


#: Define ``typer`` object that can be called from ``__main__.py``.
typer_click_object = typer.main.get_command(app)
