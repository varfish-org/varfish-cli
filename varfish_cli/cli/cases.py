"""Implementation of varfish-cli subcommand "cases *"."""

import typing
import uuid

import typer

from varfish_cli import api, common
from varfish_cli.cli.common import ListObjects, RetrieveObject
from varfish_cli.common import OutputFormat

#: Default fields for projects.
DEFAULT_FIELDS_CASE: typing.Dict[OutputFormat, typing.Optional[typing.Tuple[str, ...]]] = {
    OutputFormat.TABLE.value: ("sodar_uuid", "name", "index", "members"),
    OutputFormat.CSV.value: None,
    OutputFormat.JSON.value: None,
}

#: The ``Typer`` instance to use for the ``cases`` sub command.
app = typer.Typer(no_args_is_help=True)


@app.command("case-list")
def cli_case_list(
    ctx: typer.Context,
    project_uuid: typing.Annotated[
        uuid.UUID, typer.Argument(..., help="UUID of project to list cases for")
    ],
    output_file: typing.Annotated[
        str, typer.Option("--output-file", help="Path to file to write to")
    ] = "-",
    output_format: typing.Annotated[
        OutputFormat, typer.Option("--output-format", help="Output format")
    ] = OutputFormat.TABLE.value,
    output_delimiter: typing.Annotated[
        str, typer.Option("--output-delimiter", help="Delimiter for CSV output")
    ] = ",",
    output_fields: typing.Annotated[
        typing.Optional[typing.List[str]], typer.Option("--output-fields", help="Output fields")
    ] = None,
):
    """List all Case entries for the"""
    common_options: common.CommonOptions = ctx.obj

    list_objects = ListObjects(api.Case)
    return list_objects.run(
        common_options=common_options,
        callable=api.case_list,
        output_file=output_file,
        output_format=output_format,
        output_delimiter=output_delimiter,
        output_fields=output_fields,
        parent_uuid=project_uuid,
        default_fields=DEFAULT_FIELDS_CASE,
    )


@app.command("case-retrieve")
def cli_case_retrieve(
    ctx: typer.Context,
    object_uuid: typing.Annotated[
        uuid.UUID, typer.Argument(..., help="UUID of the object to retrieve")
    ],
    output_file: typing.Annotated[
        str, typer.Option("--output-file", help="Path to file to write to")
    ] = "-",
):
    """Retrieve Case by UUID"""
    common_options: common.CommonOptions = ctx.obj

    retrieve_object = RetrieveObject(api.Case)
    return retrieve_object.run(
        common_options=common_options,
        callable=api.case_retrieve,
        key_name="case_uuid",
        object_uuid=object_uuid,
        output_file=output_file,
    )
