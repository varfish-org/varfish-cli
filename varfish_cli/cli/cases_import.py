"""Implementation of varfish-cli subcommand "cases-import *"."""

import datetime
import os
import typing
import uuid

from google.protobuf.json_format import MessageToDict, ParseDict, ParseError
from logzero import logger
from phenopackets import Family
import typer
import yaml

from varfish_cli import api, common
from varfish_cli.cli.common import ListObjects, RetrieveObject
from varfish_cli.common import DEFAULT_PATH_VARFISHRC, OutputFormat

#: The ``Typer`` instance to use for the ``cases-import`` sub command.
app = typer.Typer(no_args_is_help=True)


@app.command("bootstrap-phenopackets")
def cli_bootstrap_phenopackets(
    ctx: typer.Context,
    project_uuid: typing.Annotated[uuid.UUID, typer.Argument(..., help="UUID of project")],
    phenopacket_file: typing.Annotated[
        str, typer.Argument(..., help="Path of phenopacket file to bootstrap")
    ],
    other_files: typing.Annotated[
        typing.List[str], typer.Argument(..., help="Paths of files to incorporate")
    ],
    target_region: typing.Annotated[
        typing.List[str], typer.Option("--target-region", help="Target region specification")
    ],
    config_path: typing.Annotated[
        str,
        typer.Option("--config-path", help="Path to configuration file", envvar="VARFISH_RC_PATH"),
    ] = DEFAULT_PATH_VARFISHRC,
):
    """Bootstrap a new or existing phenopackets YAML file"""
    # load configuration for the selected project
    logger.info("Loading configuration for project %s from %s", project_uuid, config_path)
    project_config = common.load_project_config(project_uuid, config_path=config_path)
    if not project_config:
        logger.error("No configuration found for project %s", project_uuid)
        raise typer.Exit(1)

    # split files by file type
    assert False

    # if we do not have a phenopacket file, ensure that we have a PED file
    assert False

    # load phenopacket file or create new one
    family: Family
    create_output: bool
    if os.path.exists(phenopacket_file):
        create_output = False
        with open(phenopacket_file, "rt") as inputf:
            try:
                yaml_dict = yaml.safe_load(inputf)
            except yaml.parser.ParserError as e:
                logger.error("Could not parse phenopacket YAML file: %s", e)
                raise typer.Exit(1)
        if "family" not in yaml_dict:  # pragma: no cover
            logger.error("No 'family' section found at top of phenopacket YAML file")
            raise typer.Exit(1)
        try:
            family = ParseDict(js_dict=yaml_dict["family"], message=Family())
        except ParseError as e:  # pragma: no cover
            logger.error("Could not load phenopacket data: %s", e)
            raise typer.Exit(1)
    else:
        create_output = True
        family = Family()

    # sync members in PED and phenopackets file
    assert False

    # write out phenopackets file
    if create_output:
        logger.info("Creating new phenopacket file %s", phenopacket_file)
    else:
        timestamp = datetime.datetime.now().strftime("%y%m%d-%H%M%S")
        path_bak = f"{phenopacket_file}.bak~{timestamp}"
        logger.info("Move file %s => %s", phenopacket_file, path_bak)
        os.rename(phenopacket_file, path_bak)
        logger.info("Re-creating phenopacket file %s", phenopacket_file)
    with open(phenopacket_file, "wt") as outputf:
        yaml.dump({"family": MessageToDict(family)}, outputf)

    logger.info("All done. Have a nice day!")
