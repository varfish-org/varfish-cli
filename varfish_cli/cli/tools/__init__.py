"""Implementation of varfish-cli subcommand "tools *"."""

import csv
import fnmatch
import gzip
from typing import Annotated, Dict, List

from logzero import logger
import typer

from varfish_cli.cli.tools.models import BamQc, BamQcData
from varfish_cli.config import CommonOptions
from varfish_cli.parse_ped import parse_ped

#: The ``Typer`` instance to use for the ``tools`` sub command.
app = typer.Typer(no_args_is_help=True, pretty_exceptions_enable=False)


def load_sample_data(qc_coverage_region_path: str, mapping_metrics_path: str) -> BamQcData:
    """Load data for one sample from DRAGEN QC files.

    :param qc_coverage: Path to QC coverage file
    :param mapping_metrics: Path to coverage metrics file
    :raises typer.Exit: If there is a problem with reading the input files.
    """
    fieldnames = ["_coverage_summary", "_empty", "label", "value"]
    qc_coverage_region: Dict[str, str] = {}
    with open(qc_coverage_region_path, "rt") as qc_coverage_file:
        reader = csv.DictReader(
            qc_coverage_file,
            fieldnames=fieldnames,
            delimiter=",",
        )
        qc_coverage_region = {row["label"]: row["value"] for row in reader}
    mapping_metrics: Dict[str, str] = {}
    with open(mapping_metrics_path, "rt") as coverage_metrics_file:
        reader = csv.DictReader(coverage_metrics_file, fieldnames=fieldnames, delimiter=",")
        mapping_metrics = {row["label"]: row["value"] for row in reader}

    key_aligned_in_region = "Aligned bases in QC coverage region"
    try:
        aligned_bases_in_region = int(qc_coverage_region[key_aligned_in_region])
    except (KeyError, ValueError) as e:
        logger.error(f"'{key_aligned_in_region}' not in {qc_coverage_region_path} or no int")
        raise typer.Exit() from e
    key_avg_cov_in_region = "Average alignment coverage over QC coverage region"
    try:
        avg_cov_in_region = float(qc_coverage_region[key_avg_cov_in_region])
    except (KeyError, ValueError) as e:
        logger.error(f"'{key_avg_cov_in_region}' not in {qc_coverage_region_path} or no float")
        raise typer.Exit() from e

    key_pct = "PCT of QC coverage region with coverage"
    min_cov_target: Dict[int, float] = {}
    min_cov_keys = list(
        reversed([k for k in qc_coverage_region.keys() if k.startswith(key_pct) and ": inf" in k])
    )
    for key in min_cov_keys:
        min_cov = int(key[len(key_pct) + 2 :].split(":")[0][:-1].strip())
        try:
            min_cov_target[min_cov] = float(qc_coverage_region[key])
        except (KeyError, ValueError) as e:
            logger.error(f"'{key}' not in {qc_coverage_region_path} or no float")
            raise typer.Exit() from e

    key_total_input_reads = "Total input reads"
    try:
        total_input_reads = int(mapping_metrics[key_total_input_reads])
    except (KeyError, ValueError) as e:
        logger.error(f"'{key_total_input_reads}' not in {mapping_metrics_path} or no int")
        raise typer.Exit() from e
    key_duplicate_reads = "Number of duplicate marked reads"
    try:
        duplicate_reads = int(mapping_metrics[key_duplicate_reads])
    except (KeyError, ValueError) as e:
        logger.error(f"'{key_duplicate_reads}' not in {mapping_metrics_path} or no int")
        raise typer.Exit() from e
    key_insert_length_mean = "Insert length: mean"
    try:
        insert_length_mean = float(mapping_metrics[key_insert_length_mean])
    except (KeyError, ValueError) as e:
        logger.error(f"'{key_insert_length_mean}' not in {mapping_metrics_path} or no int")
        raise typer.Exit() from e
    key_insert_length_std = "Insert length: standard deviation"
    try:
        insert_length_std = float(mapping_metrics[key_insert_length_std])
    except (KeyError, ValueError) as e:
        logger.error(f"'{key_insert_length_std}' not in {mapping_metrics_path} or no int")
        raise typer.Exit() from e

    return BamQcData(
        summary={
            "mean coverage": avg_cov_in_region,
            "total target size": int(aligned_bases_in_region / avg_cov_in_region),
        },
        min_cov_target=min_cov_target,
        bamstats={
            "sequences": total_input_reads,
            "reads duplicated": duplicate_reads,
            "insert size average": insert_length_mean,
            "insert size standard deviation": insert_length_std,
        },
    )


def load_bam_qc(
    samples: List[str],
    sample_qc_coverage_region: Dict[str, str],
    mapping_metrics: Dict[str, str],
) -> BamQc:
    """Load the BAM QC data

    :param samples: List of sample names
    :param sample_qc_coverage: Mapping from sample name to QC coverage file name
    :param mapping_metrics: Mapping from sample name to mapping metrics file name
    :return: ``BamQc`` object
    """
    return BamQc(
        sample_data={
            sample: load_sample_data(sample_qc_coverage_region[sample], mapping_metrics[sample])
            for sample in samples
        }
    )


@app.command("dragen-to-bam-qc")
def dragen_to_bam_qc(
    ctx: typer.Context,
    output_file: Annotated[str, typer.Argument(..., help="Path to 'legacy' varfish BAM QC")],
    input_files: Annotated[
        List[str],
        typer.Argument(
            ...,
            help="Path to DRAGEN QC files (*.ped, *.qc-coverage*.csv and *.mapping_metrics.csv)",
        ),
    ],
):
    """Convert DRAGEN QC files to legacy 'bam-qc' format.

    :param ctx: Typer Context
    :param output_file: Path to output file
    :param input_files: List of input files
    :raise typer.Exit: If input files are not found
    """
    common_options: CommonOptions = ctx.obj
    _ = common_options
    logger.info("Converting DRAGEN QC information")

    # Check that output path ends in .tsv or .tsv.gz.
    if not output_file.endswith((".tsv", ".tsv.gz")):
        logger.error("Output file must end in .tsv or .tsv.gz")
        raise typer.Exit()

    # Look for exactly one PED file and read it.
    logger.info("Looking for PED file ...")
    ped_file = None
    for input_file in input_files:
        if ped_file and input_file.endswith((".ped", ".ped.gz")):
            logger.error("Found more than one PED file.")
            raise typer.Exit()
        if input_file.endswith(".ped"):
            logger.info(f"... found PED file: {input_file}")
            ped_file = open(input_file, "rt")
        elif input_file.endswith(".ped.gz"):
            logger.info(f"... found PED.gz file: {input_file}")
            ped_file = gzip.open(input_file, "rt")
    if not ped_file:
        logger.error("No PED file found (looked for *.ped, *.ped.gz)")
        raise typer.Exit()

    logger.info("... PED is: \n{}")
    with ped_file:
        donors = parse_ped(ped_file)
    samples = [donor.name for donor in donors]
    logger.info(f"... donors: {donors}")

    logger.info("Looking for QC files ...")
    # Check that input paths contains one `*.qc-coverage-*` and one `*.mapping_metrics.csv`
    # file per sample:
    sample_qc_coverage = {}
    mapping_metrics = {}
    for sample in samples:
        pat_qc_coverage = f"*/{sample}.qc-coverage-*.csv"
        pat_mapping_metric = f"*/{sample}.mapping_metrics.csv"
        cand_qc_coverage = fnmatch.filter(input_files, pat_qc_coverage)
        if cand_qc_coverage:
            sample_qc_coverage[sample] = cand_qc_coverage[0]
            logger.info(f"  {sample} => {cand_qc_coverage[0]}")
        else:
            logger.error(f"  {sample} => no match for {pat_qc_coverage}")
        cand_mapping_metrics = fnmatch.filter(input_files, pat_mapping_metric)
        if cand_mapping_metrics:
            mapping_metrics[sample] = cand_mapping_metrics[0]
            logger.info(f"  {sample} => {cand_mapping_metrics[0]}")
        else:
            logger.error(f"  {sample} => no match for {mapping_metrics}")
    if len(sample_qc_coverage) != len(samples) or len(mapping_metrics) != len(samples):
        logger.error("Not all samples have QC files (see above)")
        raise typer.Exit()
    else:
        logger.info("... all samples have QC files")

    # Load data and create ``BamQc`` object.
    print(samples, sample_qc_coverage, mapping_metrics)
    bam_qc: BamQc = load_bam_qc(samples, sample_qc_coverage, mapping_metrics)

    # Write result.
    logger.debug(f"Opening output file {output_file} ...")
    if output_file.endswith(".tsv.gz"):
        output_file = gzip.open(output_file, "wt")
    else:
        output_file = open(output_file, "wt")
    with output_file:
        json_str = bam_qc.model_dump_json().replace('"', '"""')
        print(f"case_id\tset_id\tbam_stats\n.\t.\t{json_str}", file=output_file)
    logger.debug("... done writing output file.")

    logger.info("All done. Have a nice day! ðŸ˜Š")
