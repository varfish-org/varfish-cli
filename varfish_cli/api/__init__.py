"""Code for accessing VarFish Server API."""

from varfish_cli.api.case import *  # noqa: F403, F401
from varfish_cli.api.models import *  # noqa: F403, F401
from varfish_cli.api.varannos import *  # noqa: F403, F401
