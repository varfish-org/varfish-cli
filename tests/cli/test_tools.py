"""Tests for the varfish_cli.tools module."""

from pytest_mock import MockerFixture
from syrupy import SnapshotAssertion
from typer.testing import CliRunner

from tests.conftest import FakeFs
from varfish_cli.cli import app
from varfish_cli.cli.tools import load_bam_qc


def test_load_bam_qc(
    snapshot: SnapshotAssertion,
):
    samples = ["sample", "sample2"]
    cov_metrics = {
        "sample": "tests/cli/data/dragen_qc/sample.qc-coverage-region-1_coverage_metrics.csv",
        "sample2": "tests/cli/data/dragen_qc/sample2.qc-coverage-region-1_coverage_metrics.csv",
    }
    mapping_metrics = {
        "sample": "tests/cli/data/dragen_qc/sample.mapping_metrics.csv",
        "sample2": "tests/cli/data/dragen_qc/sample2.mapping_metrics.csv",
    }
    result = load_bam_qc(samples, cov_metrics, mapping_metrics)
    assert result == snapshot


def test_dragen_to_bam_qc(
    runner: CliRunner,
    fake_fs_configured: FakeFs,
    snapshot: SnapshotAssertion,
    tmpdir: str,
    mocker: MockerFixture,
):
    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    result = runner.invoke(
        app,
        [
            "--verbose",
            "tools",
            "dragen-to-bam-qc",
            f"{tmpdir}/OUT.tsv",
            "tests/cli/data/dragen_qc/sample2.mapping_metrics.csv",
            "tests/cli/data/dragen_qc/sample2.qc-coverage-region-1_coverage_metrics.csv",
            "tests/cli/data/dragen_qc/sample.mapping_metrics.csv",
            "tests/cli/data/dragen_qc/sample.qc-coverage-region-1_coverage_metrics.csv",
            "tests/cli/data/dragen_qc/samples.ped",
        ],
    )

    assert result.exit_code == 0, result.output
    assert result.output == snapshot

    with open(f"{tmpdir}/OUT.tsv", "rt") as f:
        assert f.read() == snapshot
