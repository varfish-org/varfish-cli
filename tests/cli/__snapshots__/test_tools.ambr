# serializer version: 1
# name: test_dragen_to_bam_qc
  '''
  ['sample', 'sample2'] {'sample': 'tests/cli/data/dragen_qc/sample.qc-coverage-region-1_coverage_metrics.csv', 'sample2': 'tests/cli/data/dragen_qc/sample2.qc-coverage-region-1_coverage_metrics.csv'} {'sample': 'tests/cli/data/dragen_qc/sample.mapping_metrics.csv', 'sample2': 'tests/cli/data/dragen_qc/sample2.mapping_metrics.csv'}
  
  '''
# ---
# name: test_load_bam_qc
  BamQc(sample_data={'sample': BamQcData(summary={'mean coverage': 35.44, 'total target size': 39036020}, min_cov_target={0: 100.0, 1: 99.0, 3: 98.77, 10: 98.18, 15: 97.59, 20: 95.41, 50: 0.88, 100: 0.05, 500: 0.03, 1000: 0.03, 1500: 0.03}, bamstats={'sequences': 1055691016, 'reads duplicated': 65335430, 'insert size average': 488.03, 'insert size standard deviation': 115.27}), 'sample2': BamQcData(summary={'mean coverage': 35.44, 'total target size': 39036020}, min_cov_target={0: 100.0, 1: 99.0, 3: 98.77, 10: 98.18, 15: 97.59, 20: 95.41, 50: 0.88, 100: 0.05, 500: 0.03, 1000: 0.03, 1500: 0.03}, bamstats={'sequences': 1055691016, 'reads duplicated': 65335430, 'insert size average': 488.03, 'insert size standard deviation': 115.27})})
# ---
